### ISSUE TITLE
[Rework Needed] - Gamify the health system and modularize it

### ISSUE THREAD

**Issue by MaxIsJoe** [unitystation/unitystation#9178]
After the last few health simplification PRs we have gained a boost in performance that has managed to reduce the load on the servers finally but there is still a lot more to be desired from a development and game design standpoint. As of the time of writing this issue the health script is over 1k lines of code and houses a lot of systems and features that are still too complicated or too demanding/complex for anyone to work on and some of these things are absolutely not desired from the perspective of the player or developer currently.

We need to sit down and rethink the entire system so that it is approached in a gamified manner and not a system that attempts to simulate how the biological bodies work in real life. Gamifying our systems allow developers to approach the code more easily and reduces the likelyhood of bugs happening because of the simplified approach that we should take.

This rework can be looked at after the mind rework is done, until then we must discuss and document how we want this rework to be laid out on discord in the #health channel.

Once we start development on this rework we should follow the same philosophy that we're using with Signals and the Mind Rework and make sure that everything we change is done in small chunks under multiple PRs as it lets us quickly discover game breaking bugs and take care of it before working on anything else. This is important so we don't repeat the HealthV2 incident.
Also, this rework should not be called HealthV3 and we should remove the old V1 code. One Health, Works everywhere.

## What is needed for this rework :

- Health must be divided into multiple systems that only affect general info on the main health controller script. (Modularization)
- Systems must be independent of each other and can be added/removed whenever needed. (Modularization)
- Systems must be simple and easy to approach from both the player's and developer's view. (Simple code)
- Systems must not attempt to simulate anything. (No complex systems or calculations/state checks via periodic updates)
- Systems must only be intractable via events. (Modularization)
- Systems must not contain any features that do not have a general purpose and must have a good gameplay reason for its existence.
- Health and its systems must be universal, so it can be attached to any mob, item or object.
- Document all code thoroughly and update the developer docs to include a page dedicated to how health works, how each system operates and an example for how to properly add a new system for a specie(s).

**Comment by Bod9001** [comment]
"Systems must not attempt to simulate anything. (No complex systems or calculations/state checks via periodic updates)"

that's actually the entirety of SS13 you removed there, might as well just remove atmospherics just make it so" hole in station = player get Pushed to it"


"Systems must only be intractable via events. (Modularization)"  I do get what this means?

"Health must be divided into multiple systems that only affect general info on the main health controller script. (Modularization)"

 like there is issue with Respiratory system and circulatory system, 
 
 but it's already modularised since each body part handles what it wants to do
 
 "Systems must be independent of each other and can be added/removed whenever needed. (Modularization)"
 
 circulatory system,  Could be done better, worth splitting up into separate subsystems
 
 
"Systems must be simple and easy to approach from both the player's and developer's view. (Simple code)"

It is simple

**Comment by MaxIsJoe** [comment]
> that's actually the entirety of SS13 you removed there

Simulating health is not as important as atoms because atoms actually has more uses from it being simulated. Health does not need to be simulated; it must be smoke and mirrors.
Also, we should not copy everything SS13 does because SS13 has notoriously bad design decisions.
On top of that, I don't see why are you arguing against saving performance by making our systems less demanding on the server by just simply having them not calculate stuff all the time on every body part? If we want UnityStation to be accessible one day and have the ability to support over 80+ players, we must accept that the way we're handling things currently is just terrible.

> "Systems must only be intractable via events. (Modularization)" I do get what this means?

The idea of this is that things only happens when the player interacts with the body itself, and the effects of those actions are instantaneous or gradual over-time in courtines.
Instead of checking for a new change every second, the changes happen when the player does them.

For example, instead of checking for poison every second in all body parts; the game just tells the body part it wants to affect immediately that it has been poisoned.

> but it's already modularised since each body part handles what it wants to do

Not modularized enough because you're still dependent on other systems, and I don't like how every body part has a lot of logic and calculations that need to be checked. I also don't like how if you strip away one system from something, the entire body just breaks with some of them throwing NREs in some cases.

> Circulatory system, Could be done better, worth splitting up into separate subsystems

Glad we agree on that. Still, it's better to discuss how we want to design the circulatory system on discord, just like how we discussed the thief role until we come up with a satisfying system that is easy to approach from the developer's side and is less prone to bugs.

> It is simple

No, It is not.
Proof? Look at how gigantic our health scripts are and how complex some of them are.
I tried fixing multiple bugs in health before, but I simply couldn't because I kept jumping between multiple functions and entire scripts to just understand what the hell is even going on, and there are some useless features (such as fat) that just need to removed and either receive a gamified treatment to their design or simply put back on the drawing board because of how useless or bloated they are currently.

**Comment by MaxIsJoe** [comment]
Because I know the argument of simulating health will be brought up again, Let me explain to you why simulating health is a terrible idea while atmospherics are 100% fine:

When atmos is simulated, you see the effects of its simulation in-game instantly with visual queues to wind and things being pushed around + it damaging characters when they're not wearing a gas mask to inhale the appropriate gases that the character needs to survive.
The simulation and its effects are always seen by the player and communicated clearly via multipe means, such as text, sound and gameplay features.

For health however:

- You do not see your blood flowing through your entire body and there is no gameplay use to blood flow so it's 100% useless.
- You do not have a general purpose to simulate heart beats and gameplay wise you never even touch that side of the heart organ so it's existence is 100% useless.
- You do not see fat acculemating on your body and I don't see us turning into a furry inflation fetishist remake to fully simulate that, so it's 100% useless because you never see or interact with it in-game and the idea of moving fat from one body to another to make people suddenly be not hungry via surgery is absolutely stupid when IV drips are a thing.
- body part modifiers are only used on two limb types and everywhere else their effect is not seen or felt by players because their only use case is self-heal and most players don't know that and it STILL doesn't need to be simulated as the metabolism system can just handle that for you in a gamified way.
- digesting reagents does NOT need to be simulated, as this can get out of hand quickly if there are a lot of reagents in lots of stomachs which is just server power being wasted even if our chemistry system is highly optimized.. again, just because we have the power to do something like this doesn't mean we should do it as that is processing power that can go elsewhere every update. You never see what happens in your stomach and you don't know anything about it other than the static hunger bar that updates once every couple of minutes after you've eaten something.. We're also never going to add vore so simulating this is useless unless you want to pander to vore fetishists escaping from skyrat to make a fully simulated digestion system on unitystation.

And I can go on and on.
TLDR : Just because we can simulate the body does not mean we should as it's processing power that could go elsewhere and simulating the body is just a big mess that is hard to maintain and can break easily. Almost all the simulation stuff that happens in health is never seen by the player and has no important or fun gameplay uses what-so-ever.

**Comment by Bod9001** [comment]
fat it is not bloated, it's very simple,

if blood contain too much nutriment, Absorb the nutriment
if blood contain too little nutriment, release, 

if fat full Spawn new fat

**Comment by Bod9001** [comment]
the simulation aspect is fine it's optimally Speed

**Comment by MaxIsJoe** [comment]
> fat it is not bloated, it's very simple,
> 
> if blood contain too much nutriment, Absorb the nutriment if blood contain too little nutriment, release,
> 
> if fat full Spawn new fat

Again, we do not need this to be simulated.
This can just be a simple number that represents how hungry the player is and how slow they can get when their stomach is full.
Fat is an entire organ and class that only serves one purpose in its entire existence, and when it breaks, other stuff break with it.

**Comment by PetMudstone** [comment]
Yeah I have to agree with Max. I think Jack said once that stuff should only be as complex as it needs to be to be to provide benefit to the player. If you get the same result with a more complex system as you do with a simple system, then there's no benefit to doing a more complex system.

Just because SS13 is a simulation doesn't mean that you need to simulate literally everything. Not everything that can be simulated is worth simulating. /tg/station's experiement with hygiene is one example of this. Simulating personal uncleanliness and forcing people to take showers or face penalties isn't engaging gameplay.

Again, if you're simulating stuff you need an actual reason why it's worth simulating. What benefit does modeling heartbeats and bloodflow throughout the body give to players? How does it improve their experience? You bring up atmospherics but its simulation has many very visible uses across the entire game and its existence interacts with many other systems as well. Bombmaking, construction, plasmaflooding, power generation, basic survival etc. What does simulating fat the way we do add to the players' experience?

**Comment by MaxIsJoe** [comment]
![image](https://c.tenor.com/pyGugHcm4XcAAAAC/michael-scott-the-office.gif)

**Comment by Bod9001** [comment]
no, is a simple fact that it doesn't have any data redundancy, 
you remove that fat ,  and you remove the debuffs with it, simple as, you don't have to worry about updating some variable on the player, because it's intrinsic to the fact that it exists there current system could be worked on a bit but it's good

**Comment by PetMudstone** [comment]
Is it really bad to have nutrition be a variable on the player themself? After all, we were already able to roughly model the effects of hunger before the health refactor without fat being a physical "organ". /tg/station does too, with added functionality like overeating and dietary preferences. You can get "fat" in /tg/ with very visible detriments without fat actually being modelled, just a mere stage after a variable gets high enough. You can even do surgeries on people once they get fat enough to reduce it.

Again, the issue is that stuff is being modeled and processed when you can get a near identical if not actually identical end result in a simpler manner that's easier to comprehend for both developers and players. Please explain again why we need a fat organ rather than a simple variable.

**Comment by MaxIsJoe** [comment]
> You Remove fat and you remove the debuff with it.

You also remove fats at the start and a math error happens
![image](https://user-images.githubusercontent.com/34368774/187575679-2dcd1ee3-6fe3-4374-b5c3-ad60e86fdb55.png)

Again, this could be simplified by simply not constantly spawning an organ that only serves one very small and insignificant purpose and having it's debuffs represented normally with numbers in the metabolism system. Not all species would have fats as well in the first place, so this is hard-coded into the system to force all characters to have fats even if some don't make sense to have it like robot based species.
This still does not dismiss the fact that we need to make this system independent so we don't have ugly things like initializing hunger on one script and calculating nuritiement on another, everything should be sandboxed in one area, so the code becomes much cleaner and easier to work with.

**Comment by corp-0** [comment]
I agree we need to simplify health, it is not secret how much I don't like how every bodypart has hunger. 

That said, I don't agree at all with statements like "``Systems must not attempt to simulate anything. (No complex systems or calculations/state checks via periodic updates)``". That doesn't make any sense.

**Comment by MaxIsJoe** [comment]
I should re-word this then.
We need to make our systems not hog the update loop with constant for-each checks and calculations, If something can be emulated via events then we do not need to constantly check for it in an update loop.
The system also does not need to behave with a degree of realism nor does it need to have a simulated behavior via code. We can fake the simulation by gamifying the systems, but we absolutely do not need to model a complex system it in-code that will not have any noticeable gameplay uses that will also just be a pain to maintain for developers.
Again, we're trying to create a system that's just smoke and mirrors; in terms of gameplay we can claim that we simulate something, but behind the scenes we shouldn't do that.

**Comment by AtnerNT** [comment]
Just commenting to express my agreement with the simplifying sentiment.

**Comment by MaxIsJoe** [comment]
Closing this because honestly we're way past doing any more changes to HealthV2, and it's been somewhat stable for a while now.
